### ISSUE TITLE
[apex] UnusedLocalVariable: False positive in dynamic SOQL

### ISSUE THREAD

**Issue by parsam97** [pmd/pmd#6448]
**Affects PMD Version:** 7.15.0

**Rule:** [UnusedLocalVariable](https://docs.pmd-code.org/latest/pmd_rules_apex_bestpractices.html#unusedlocalvariable)

**Description:** The syntax for referencing local variables in dynamic SOQL is to do `'...WHERE Id = :var'` within a string. To reference local variables in dynamic SOQL, we must do it that way and cannot do something like:
`String query = 'SELECT Id FROM Account WHERE Id IN ' + idSet';`

**Code Sample demonstrating the issue:**

```apex
Set<Id> idSet = new Set<Id>{ accountId };
String query = 'SELECT Id FROM Account WHERE Id IN :idSet';
```

OR

```apex
Set<Id> idSet = new Set<Id>{ accountId };
fflib_QueryFactory qf = newQueryFactory()
    .setCondition('Id IN :idSet');
```

**Expected outcome:**

PMD reports a violation at line 1 (both examples), but that's wrong. That's a false positive. `idSet` is being used in dynamic SOQL.

I imagine this problem would also occur for dynamic SOSL:

```apex
String term = 'search-term';
String s = 'FIND :term IN ALL FIELDS RETURNING Account(Id)';
```

**Running PMD through:** *[Apex PMD vscode extension](https://github.com/ChuckJonas/vscode-apex-pmd)*

**Comment by LINTHESHWARS** [comment]
@parsam97 Is this issue still open, could also you provide some more details on replicating the issue?

**Comment by parsam97** [comment]
@LINTHESHWARS  Sure, just let me know what more detail I could give? Code demonstrating the issue was already given.

**Comment by Lintheshwar** [comment]
@parsam97 Could you provide detailed steps on reproducing the issue? And if possible screenshot of the error that is occurring?

**Comment by adangel** [comment]
The rule UnusedLocalVariable supports SOQL queries. But in your example, we just see a String - so we can't know, that this is not just a string but actually a query.

This issue is very similar to #5000.

Have you tried this code:

```apex
Set<Id> idSet = new Set<Id>{ accountId };
String query = 'SELECT Id FROM Account WHERE Id IN :idSet';
Database.query(query);
```

This _should not_ report idSet as unused, as we should detect that query is used in the method call for Database.query.

The important part is, that the query String is used directly in a `Database.query`, `Database.getQueryLocator` or `Database.countQuery`. That is how PMD knows, that the string is not just a string, but a database query.

What is `newQueryFactory()`. Where is it defined? Is it a Apex built-in function? Or is it provided by a library?

**Comment by adangel** [comment]
Is it maybe this one? https://github.com/apex-enterprise-patterns/fflib-apex-common/blob/bee8e18602081fbf2f5180270d027ea782707b2a/sfdx-source/apex-common/main/classes/fflib_QueryFactory.cls#L54

**Comment by parsam97** [comment]
@adangel Thanks for your response. Firstly I hadn't noticed that PMD notices when doing this:
```
String someName;
String query = 'SELECT Id FROM Account WHERE Name = :someName';
Database.query(query);
```

That's cool that there's no violation there. But as you noticed, it is very common practice to use some form of query builder to build queries, and a selector to execute those queries. You've got the right link to the one I use.

So my `CaseSelector` for instance can look like this:
```
public with sharing class CaseSelector
    extends ApplicationSObjectSelector
    implements ICaseSelector
{
    public List<Case> selectById(Set<Id> caseIds)
    {
        if (caseIds == null || caseIds.isEmpty())
            return new List<Case>();

        Set<Id> caseIds2 = caseIds.remove(0); // <-- violation here

        fflib_QueryFactory qf = buildQueryFactory()
            .setCondition('Id IN :caseIds2');

        return Database.query(qf.toSOQL());
    }
}
```

I put `caseIds2` just as an example to show that there is a violation on the line that declares `caseIds2`. In this code example, `Database.query()` is being called but since the reference is buried in the `qf` variable and only exposed when doing `qf.toSOQL()`, I would imagine it would be very costly and near impossible to fix this, unless you change the rule entirely to be based on something else like the presence of the `:` before it.

**Comment by parsam97** [comment]
In the issue you linked ([#5000](https://github.com/pmd/pmd/issues/5000)), the code is passing the query string directly to the first parameter like so: `Database.query(dynamicQueryString)`
But this breaks when you have some form of query builder in place that builds the query for you.
